site/src/components/WorkspaceScheduleButton/WorkspaceScheduleButton.tsx

```diff
@@ -0,0 +1,249 @@
+  autoStopDisplay: (workspace: Workspace): string => {
+    const deadline = dayjs(workspace.latest_build.deadline).utc()
+    // a manual shutdown has a deadline of '"0001-01-01T00:00:00Z"'
+    // SEE: #1834
+    const hasDeadline = deadline.year() > 1
+    const ttl = workspace.ttl_ms
+
+    if (isWorkspaceOn(workspace) && hasDeadline) {
+      // Workspace is on --> derive from latest_build.deadline. Note that the
+      // user may modify their workspace object (ttl) while the workspace is
+      // running and depending on system semantics, the deadline may still
+      // represent the previously defined ttl. Thus, we always derive from the
+      // deadline as the source of truth.
+      const now = dayjs().utc()
+      if (now.isAfter(deadline)) {
+        return Language.workspaceShuttingDownLabel
+      } else {
+        return deadline.tz(dayjs.tz.guess()).format("MMM D, YYYY h:mm A")
+      }
+    } else if (!ttl || ttl < 1) {
+      // If the workspace is not on, and the ttl is 0 or undefined, then the
+      // workspace is set to manually shutdown.
+      return "Manual"
+    } else {
+      // The workspace has a ttl set, but is either in an unknown state or is
+      // not running. Therefore, we derive from workspace.ttl.
+      const duration = dayjs.duration(ttl, "milliseconds")
+      return `${duration.humanize()} after start`
+    }
+  },
+  editScheduleLink: "Edit schedule",
+  editDeadlineMinus: "Subtract one hour",
+  editDeadlinePlus: "Add one hour",
+  scheduleHeader: (workspace: Workspace): string => {
+    const tz = workspace.autostart_schedule
+      ? extractTimezone(workspace.autostart_schedule)
+      : dayjs.tz.guess()
+    return `Schedule (${tz})`
+  },
+}
+
+export const shouldDisplayPlusMinus = (workspace: Workspace): boolean => {
+  if (!isWorkspaceOn(workspace)) {
+    return false
+  }
+  const deadline = dayjs(workspace.latest_build.deadline).utc()
+  return deadline.year() > 1
+}
+
+export const deadlineMinusDisabled = (workspace: Workspace, now: dayjs.Dayjs): boolean => {
+  const delta = dayjs(workspace.latest_build.deadline).diff(now)
+  return delta <= 30 * 60 * 1000 // 30 minutes
+}
+
+export const deadlinePlusDisabled = (workspace: Workspace, now: dayjs.Dayjs): boolean => {
+  const delta = dayjs(workspace.latest_build.deadline).diff(now)
+  return delta >= 24 * 60 * 60 * 1000 // 24 hours
+}
+
+const WorkspaceScheduleLabel: React.FC<{ workspace: Workspace }> = ({ workspace }) => {
+  const styles = useStyles()
+
+  if (isWorkspaceOn(workspace)) {
+    const stopLabel = Language.autoStopDisplay(workspace)
+    const isShuttingDown = stopLabel === Language.workspaceShuttingDownLabel
+
+    // If it is shutting down, we don't need to display the auto stop label
+    return (
+      <span className={styles.labelText}>
+        {!isShuttingDown && (
+          <strong className={styles.labelStrong}>{Language.autoStopLabel}</strong>
+        )}
+        {stopLabel}
+      </span>
+    )
+  }
+
```

Comment:
❤️  I really love the guard statement above (the early return) and then the default return down below. This is a more readable format for me.